src/cargo/ops/cargo_compile/mod.rs

```diff
@@ -542,6 +589,52 @@ where `<compatible-ver>` is the latest version supporting rustc {rustc_version}"
     Ok(bcx)
 }
 
+fn warn_on_unexpected_cfgs(
+    gctx: &GlobalContext,
+    check_cfg: &CheckCfg,
+    cfg_expr: &CfgExpr,
+    path: Option<&Path>,
+    suffix: &str,
+) -> CargoResult<()> {
+    let prefix = if let Some(path) = path {
+        format!("{path}: ", path = path.display())
+    } else {
+        "".to_string()
+    };
+
+    cfg_expr.fold(&|cfg| -> CargoResult<()> {
+        let (name, value) = match cfg {
+            Cfg::Name(name) => (name, None),
+            Cfg::KeyPair(name, value) => (name, Some(value.to_string())),
+        };
+
+        match check_cfg.expecteds.get(name) {
+            Some(ExpectedValues::Some(values)) if !values.contains(&value) => {
+                let value = if let Some(ref value) = value {
+                    value
+                } else {
+                    "(none)"
+                };
+                gctx.shell().warn(format!(
+                    "{prefix}unexpected `cfg` condition value: `{value}` for `{cfg}` in `[target.'cfg({cfg_expr})'{suffix}]`"
+                ))?;
+            }
+            None => {
+                gctx.shell().warn(format!(
+                    "{prefix}unexpected `cfg` condition name: `{name}`{cfg} in `[target.'cfg({cfg_expr})'{suffix}]`",
+                    cfg = if value.is_some() {
+                        format!(" for `{cfg}`")
+                    } else {
+                        format!("")
+                    }
+                ))?;
+            }
```

Comment:
As this aligns closely with our new linting system, I'm assuming @Muscraft would prefer these to be using annotate-snippets

Reply:
The last revision of this PR is now using annotate-snippets.